Feature/review

nayoonk928 opened the pull request:
### 변경사항
**AS-IS**

* 사용자가 특정 컨텐츠에 리뷰 등록/수정/삭제하는 기능 구현

**TO-BE**

* 다른 사용자의 리뷰에 추천/비추천/신고하기 기능

### 테스트
- [x] 테스트 코드
- [ ] API 테스트

nayoonk928 committed:
feat: Implement review add/update/delete feature

- 인증된 사용자가 컨텐츠에 리뷰를 등록/수정/삭제하는 기능 구현

nayoonk928 committed:
test: Add test code for user add/update/delete review

mingming-mentor commented on src/main/java/com/personal/contentservice/dto/review/ReviewAddDto.java:
```diff
@@ -0,0 +1,25 @@
+  private String comment;
+
+  @NotNull
+  private Double rating;
```
@NotNull이라면 double로 해주시면 좋을 것 같습니다

mingming-mentor commented on src/main/java/com/personal/contentservice/service/impl/ReviewServiceImpl.java:
```diff
@@ -0,0 +1,115 @@
+  }
+
+  @Override
+  public String updateReview(Authentication authentication, ReviewUpdateDto request) {
+    User user = UserAuthenticationUtils.getUser(authentication);
+    Content content = getContentByIdAndMediaType(request.getContentKey());
+
+    Review review = reviewRepository.findByUserAndContent(user, content)
+        .orElseThrow(() -> new CustomException(REVIEW_NOT_FOUND));
+
+    if (request.getComment() != null) {
+      review.setComment(request.getComment());
+    }
+
+    if (request.getRating() != null) {
+      review.setRating(request.getRating());
+    }
+
+    reviewRepository.save(review);
+    calculateAverageRating(content);
+    return "리뷰가 수정되었습니다.";
+  }
```
update 메소드에서 변경이 있을 경우 데이터를 넣어주는 형식으로 보여요. 이렇게 되면 null로 업데이트를 하고자 하는 경우 적용이 어렵습니다. 그래서 보통 update method에서는 client에서 update하고자하는 형태의 snapshot을 보내줍니다. 
예) 기존 rating = 1, comment = "comment" 를 rating=2로 변경하고자 한다면 method에 rating값만 넣어주는 것이 아닌 rating=2, comment="comment" 로 넣어줍니다.
대신 서버에서 현재 데이터베이스에 들어있는 값과 같다면 별도로 patching을 해주지 않는 형태로 하는 것이죠. 
이 방식도 한번 고려해보시면 좋을 것 같아요.

mingming-mentor commented on src/main/java/com/personal/contentservice/service/impl/ReviewServiceImpl.java:
```diff
@@ -0,0 +1,115 @@
+    return content;
+  }
+
+  private void calculateAverageRating(Content content) {
```
method에 save까지 들어가면 좋을 것 같아요. 현재 메소드명으로는 평균 rating을 계산하여 반환해주는 역할로만 보이네요

mingming-mentor commented on src/main/java/com/personal/contentservice/service/impl/ReviewServiceImpl.java:
```diff
@@ -0,0 +1,115 @@
+  }
+
+  private void calculateAverageRating(Content content) {
+    List<Review> reviews = reviewRepository.findAllByContent(content);
```
이 부분 캐싱을 이용하면 좋을 것 같습니다. 리뷰 등록시마다 많은 쿼리가 발생될듯해요.

mingming-mentor commented on src/main/java/com/personal/contentservice/service/impl/ReviewServiceImpl.java:
```diff
@@ -0,0 +1,115 @@
+      double totalRating = 0.0;
+      for (Review review : reviews) {
+        totalRating += review.getRating();
+      }
```
그리고 이 부분 연산도 꽤 걸릴듯하여 캐싱을 하지 않는다면 쿼리로 연산까지 하여 반환하면 좋을 것 같아요
avg를 사용하시면 될듯합니다

mingming-mentor reviewed (approved):
리뷰드렸습니다! 수고 많으셨어요

nayoonk928 committed:
refactor: Refactor ReviewServiceImpl

- deleteReview 시 lock을 통해 추천, 비추천, 신고 등의 행위를 할 수 없도록 처리
- saveCalculateAverageRating 메서드를 쿼리로 연산까지 하여 반환하는 것으로 수정
